src/ElmFormat/Render/Box.hs

```diff
@@ -2120,6 +2072,104 @@ formatComment comment =
         AST.CommentTrickBlock c ->
             mustBreak $ row [ punc "{--", literal c, punc "-}" ]
 
+formatIfExpression ::
+    ElmVersion
+    -> ImportInfo
+    -> AST.Expression.IfClause
+    -> [(AST.Comments, AST.Expression.IfClause)]
+    -> (AST.Comments, AST.Expression.Expr)
+    -> AST.Multiline
+    -> Box
+formatIfExpression elmVersion importInfo if' elseifs (elsComments, els) multiline =
+    let
+        (cond, (AST.Commented preThen thenExpr postThen)) =
+            if'
+
+        then' =
+            concat $
+                [ Maybe.maybeToList $ formatComments preThen
+                , [ formatExpression elmVersion importInfo SyntaxSeparated thenExpr ]
+                , Maybe.maybeToList $ formatComments postThen
+                ]
+
+        else' =
+            (Maybe.maybeToList $ formatComments elsComments)
+                ++ [ formatExpression elmVersion importInfo SyntaxSeparated els ]
+    in
+    case
+        ( multiline
+        , formatCommentedExpression elmVersion importInfo SyntaxSeparated cond
+        , allSingles then'
+        , elseifs
+        , allSingles else'
+        )
+    of
+        (AST.JoinAll, SingleLine cond', Right singleThens, [], Right singleElses)  ->
+            line $ row $ concat
+                [ [ keyword "if", space, cond', space, keyword "then", space ]
+                , List.intersperse space singleThens
+                , [ space, keyword "else", space ]
+                , List.intersperse space singleElses
+                ]
+
+        (_, ifCond, thenBody, _, elseBody) ->
+            formatIfExpressionMultiline elmVersion importInfo ifCond thenBody elseifs elseBody
```

Comment:
The way these two functions work together seems a bit complicated.  The ideal is that these functions basically have two phases: 1) turning all the stuff into boxes and 2) assembling the boxes based on the multiline features.  Here we turn just enough into boxes to decide if we should call the other function, which makes it a bit harder to follow the full logic.

Though looking at the code for a few minutes, I don't have any quick suggestions of how to refactor this, so if you don't have time to give a try to simplifying this more, that's fine.  (A lot of the existing code in here is way more confusing than this)

Reply:
I think it should be easier to follow the logic now.